keps/sig-network/0015-20180614-SCTP-support.md

```diff
@@ -0,0 +1,258 @@
+      * [Story 2](#story-2)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+## Summary
+
+The goal of the SCTP support feature is to enable the usage of the SCTP protocol in Kubernetes [Service][],  [NetworkPolicy][], and [ContainerPort][]as an additional protocol value option beside the current TCP and UDP options.
+SCTP is an IETF protocol specified in [RFC4960][], and it is used widely in telecommunications network stacks.
+Once SCTP support is added as a new protocol option those applications that require SCTP as L4 protocol on their interfaces can be deployed on Kubernetes clusters on a more straightforward way. For example they can use the native kube-dns based service discovery, and their communication can be controlled on the native NetworkPolicy way.
+
+[Service]: https://kubernetes.io/docs/concepts/services-networking/service/
+[NetworkPolicy]: 
+https://kubernetes.io/docs/concepts/services-networking/network-policies/
+[ContainerPort]:https://kubernetes.io/docs/concepts/services-networking/connect-applications-service/#exposing-pods-to-the-cluster
+[RFC4960]: https://tools.ietf.org/html/rfc4960
+
+
+## Motivation
+
+SCTP is a widely used protocol in telecommunications. It would ease the management and execution of telecommunication applications on Kubernetes if SCTP were added as a protocol option to Kubernetes. 
+
+### Goals
+
+Add SCTP support to Kubernetes ContainerPort, Service and NetworkPolicy, so applications running in pods can use the native kube-dns based service discovery for SCTP based services, and their communication can be controlled via the native NetworkPolicy way.
+
+### Non-Goals
+
+It is not a goal here to add SCTP support to load balancers that are provided by cloud providers. 
+
+It is not a goal to support multi-homed SCTP associations. Such a support also depends on the ability to manage multiple IP addresses for a pod, and in the case of Services with ClusterIP or NodePort the support of multi-homed assocations would also require the support of NAT for multihomed associations in iptables/ipvs.
+
+It is not a goal to support SCTP as protocol value for the container's HostPort. The reason: [the usage of HostPort is not recommended by Kubernetes][], and to ensure proper interworking of HostPort with userspace SCTP stacks (see below) would require an additional kubelet/kubenet configuration option. In order to keep the complexity and impact of the introduction of SCTP on a lower level we do not plan to support SCTP as new protocol value for HostPort.
+
+[the usage of HostPort is not recommended by Kubernetes]:https://kubernetes.io/docs/concepts/configuration/overview/#services
```

Comment:
I don't see why not, I guess.

Reply:
OK. The reason I proposed this way was, that listening on HostPorts (in order to reserve the port) is not handled by the kube-proxy but by the kubelet (kubelet -> Docker Service -> Kubenet plugin -> hostportmanager). That is, in order to have the same "userspace sctp config" based listening on HostPorts it would require another config parameter for another component - i.e. the user shall configure the userspace sctp support in 2 different places. And because I understood the cited k8s documentation so, that the usage of HostPorts is not recommended, I thought that then I can skip the support of SCTP for HostPorts. But of course, if the decision is to have SCTP support for HostPorts, too, then I do it.

Reply:
In truth, I am not sure how valuable kubelet's logic is here.  ISTR there was/is a bug in it that it doesn't re-open ports if Kubelet restarts, and nobody has complained about it.

Maybe we should just remove that logic altogether?  We could start by skipping it for SCTP and have a separate discussion about removal...  Just make sure it is well commented why we do it for everything except SCTP.

Reply:
Fine for me.